### PR TITLE
Detect version of the key value store in use and react appropriately

### PR DESCRIPTION
**Pull request by eugene-davis**
Handle the difference between V1 and V2 responses of the key value secret store.

Still needs:

- [x]  Updates to tests
- [x] Update code so detection is in its own function

### CODE REVIEW

File: vault_anyconfig/vault_anyconfig.py [hunk 1]
```diff
@@ -163,7 +163,15 @@ def save_file_from_vault(self, file_path, secret_path, secret_key):
             - secret_path: secret's path in Vault
             - secret_key: key in the secret in Vault to access
         """
-        secret_file_string = self.read(secret_path)["data"][secret_key]
+        # Handle v1 vs v2 of the key-value secret store, a version isn't returned in the response so check the shape of the data
+        # v1: https://www.vaultproject.io/api/secret/kv/kv-v1.html#sample-response-1
+        # v2: https://www.vaultproject.io/api/secret/kv/kv-v2.html
```
Comment (martinm82) [inline vault_anyconfig/vault_anyconfig.py]: That link points as well to the response anchor: https://www.vaultproject.io/api/secret/kv/kv-v2.html#sample-response

Comment (eugene-davis) [reply vault_anyconfig/vault_anyconfig.py]: Good catch, fixed.

File: vault_anyconfig/vault_anyconfig.py [hunk 2]
```diff
@@ -90,7 +90,8 @@ def auth_from_file(self, vault_creds_file):
     def dump(self, data, out, process_secret_files=False, **args):
         """
         First updates the provided dictionary with keys from the Vault, then calls anyconfig to dump out a configuration file.
-        See https://python-anyconfig.readthedocs.io/en/latest/api/anyconfig.api.html#anyconfig.api.dump for detailed invocation options.
+        #anyconfig.api.dump for detailed invocation options.
```
Comment (martinm82) [inline vault_anyconfig/vault_anyconfig.py]: Something happened with this link.

File: vault_anyconfig/vault_anyconfig.py [hunk 3]
```diff
@@ -274,6 +282,56 @@ def __process_vault_files(self, config):
...
+        """
+        Checks if the response is from the key value v1 secret engine.
+        See https://www.vaultproject.io/api/secret/kv/kv-v1.html#sample-response-1
+        Args:
+            - read_response: repsonse from HVAC read function
+            - secret_key: secret key being retrieved
+        Returns:
+            Bool
+        """
+        if secret_key in read_response.get('data', {}):
```
Comment (martinm82) [inline vault_anyconfig/vault_anyconfig.py]: Should be doable in one line: `return secret_key in read_response.get('data', {})`

File: vault_anyconfig/vault_anyconfig.py [hunk 4]
```diff
@@ -274,6 +282,56 @@ def __process_vault_files(self, config):
 
         return
 
+    @classmethod
+    def __process_response(cls, read_response, secret_key):
+        """
+        Detects the secret engine returning a secret (currently *only* supports key-value versions 1 and 2) and returns the requested key from the
+        secret.
+        Args:
+            - read_response: repsonse from HVAC read function
```
Comment (martinm82) [inline vault_anyconfig/vault_anyconfig.py]: typo: repsonse vs. response

File: vault_anyconfig/vault_anyconfig.py [hunk 5]
```diff
@@ -274,6 +282,56 @@ def __process_vault_files(self, config):
...
+                "Invalid response recieved. Possibly due to an unsupported secrets engine, vault-anyconfig currently only supports kv1 and kv2.")
+        return secret_string
+
+    @staticmethod
+    def __is_key_value_v1(read_response, secret_key):
+        """
+        Checks if the response is from the key value v1 secret engine.
+        See https://www.vaultproject.io/api/secret/kv/kv-v1.html#sample-response-1
+        Args:
+            - read_response: repsonse from HVAC read function
```
Comment (martinm82) [inline vault_anyconfig/vault_anyconfig.py]: Typo: repsonse vs. response

See as well in the v2 function

File: vault_anyconfig/vault_anyconfig.py [hunk 6]
```diff
@@ -274,6 +282,56 @@ def __process_vault_files(self, config):
...
+            Bool
+        """
+        if secret_key in read_response.get('data', {}):
+            return True
+        return False
+
+    @staticmethod
+    def __is_key_value_v2(read_response, secret_key):
+        """
+        Checks if the response is from the key value v1 secret engine
```
Comment (martinm82) [inline vault_anyconfig/vault_anyconfig.py]: v1 vs. v2

File: vault_anyconfig/vault_anyconfig.py (hunk 2 above)
Comment (eugene-davis) [reply vault_anyconfig/vault_anyconfig.py]: Somehow my autoformatter got confused I think, since it doesn't seem to get repeated. Its reverted now.

File: vault_anyconfig/vault_anyconfig.py [hunk 7]
```diff
@@ -274,6 +278,56 @@ def __process_vault_files(self, config):
...
+        """
+        Detects the secret engine returning a secret (currently *only* supports key-value versions 1 and 2) and returns the requested key from the
+        secret.
+        Args:
+            - read_response: repsonse from HVAC read function
+            - secret_key: secret key being retrieved
+        Returns:
+            secret string
+        """
+        if cls.__is_key_value_v1(read_response, secret_key):
```
Comment (mangokingTW) [inline vault_anyconfig/vault_anyconfig.py]: `__is_key_value_v1` always returns true, because v1 and v2 both start with the 'data' block. 
And if the key is 'data' in the v1 engine, `__is_key_value_v2` will take it as v2.

Comment (eugene-davis) [reply vault_anyconfig/vault_anyconfig.py]: Yes this is a good catch, I will address this.

Comment (eugene-davis) [reply vault_anyconfig/vault_anyconfig.py]: This should now be resolved by also confirming the presence or absence of the metadata key.

File: vault_anyconfig/vault_anyconfig.py [hunk 8]
```diff
@@ -309,7 +309,7 @@ def __is_key_value_v1(read_response, secret_key):
         Returns:
             Bool
         """
-        return secret_key in read_response.get('data', {})
+        return secret_key in read_response.get('data', {}) and 'metadata' not in read_response.get('data', {})
```
Comment (mangokingTW) [inline vault_anyconfig/vault_anyconfig.py]: If the name of a key is 'metadata' in the v1 engine, this condition will return false.

Comment (eugene-davis) [reply vault_anyconfig/vault_anyconfig.py]: Yep, didn't even think about it.
Let me do some digging and see if I can find a stronger alternative detection method.

Comment (mangokingTW) [reply vault_anyconfig/vault_anyconfig.py]: I think you can try this way:
1. read_response.get('data', {}) and len(read_response) > 1 is v1
2. read_response.get('data', {}) and len(read_response) == 1 is v2

Comment (eugene-davis) [reply vault_anyconfig/vault_anyconfig.py]: Adding the approaches I'm looking at:
1. Check the mount type via the API, which is how they do it in the Vault CLI client (https://github.com/hashicorp/vault/blob/0c9b0117093df05bd7539325c500ec3fbdff982d/command/kv_helpers.go#L44). See https://www.vaultproject.io/api/system/mounts.html
2. Require explicit definition of the secret engine in the `vault_secrets` section.
3. Check for `/data/` in secret path, and document that it cannot be used in a v1 secret path
4. Document the exclusion of certain key values (such as data and metadata).
5. Increase check complexity (probably just making sure the `secret_key` isn't one of the keys I am checking for the presence of but I will check for other edge cases).

I'm excluding most of them for the following reasons:
1. Requires permissions on `/sys/mounts`, which reduces security and makes configuration more complicated. Their client isn't being used in software but rather as an admin tool.
2. Makes the file format too complicated.
3. Totally disallowing a mount path seems onerous and goes against the ideal of keeping the configuration required on Vault as minimal as possible.
4. Better than 3, but still requires extra complexity in configuring Vault. I will fallback to this approach if needed.

Comment (mangokingTW) [reply vault_anyconfig/vault_anyconfig.py]: Another possible way I found:
`isinstance(read_response.get('data', {}).get('metadata', ''), dict)`
v1 should return false.
v2 should return true.

Comment (eugene-davis) [reply vault_anyconfig/vault_anyconfig.py]: Ok, I updated with:
v1 as `isinstance(read_response.get('data', {}).get(secret_key, {}), str)`
v2 as `isinstance(read_response.get('data', {}).get('data', ''), dict)`

Assuming I also correctly implemented the tests, then this should cover at least the 'data' and 'metadata' as keys edge cases.

**Comment by eugene-davis** [comment]
I will release a new minor version once I have gotten a review from one of the required reviewers.

File: README.md [hunk 9]
```diff
@@ -15,6 +15,15 @@ Extends the [HVAC Hashicorp Vault Client](https://github.com/hvac/hvac) with the
...
 file with all details populated save for secrets, and then access Hashicorp Vault to load the secrets into the in-memory dictionary.
 
+## Supported Secret Engines
+
+Currently vault-anyconfig **only** supports version 1 and 2 of the key value store.
+
+### kv2 Limitations
+
+* vault-anyconfig **only** will read the latest version of a secret to maintain simplicity in the configuration file
+* you must add `data` after the mointpoint for a kv2 secret, e.g. `secret/data/example-secret` due to limitations in the HVAC client
```
Comment (martinm82) [inline README.md]: typo: mointpoint vs. mountpoint

File: README.md [hunk 10]
```diff
@@ -147,6 +156,9 @@ file that looks like:
 }
 ```
 
+**Key-Value Store V2 Limitation**: You must include `data` after the mountpoint, for example, `secret/mysql/customer` should be
```
Comment (martinm82) [inline README.md]: Can't you point to `# kv2 Limitation' ? What you write here is the same as in that section

Comment (martinm82) [reply README.md]: Ok, I see you did that as well for `vault_files`. I guess you did that to avoid the situation where one hasn't read the whole doc, right?

File: tests/unit/file_handling_tests/test_fuzz_secret_file.py [hunk 11]
```diff
@@ -22,7 +22,8 @@
         min_size=1,
         alphabet=strat.characters(
             blacklist_categories=("C"),
-            blacklist_characters=".",  # Since we separately specify the key, we cannot include "." in the secret path
+            # Since we separately specify the key, we cannot include "." in the secret path
```
Comment (martinm82) [inline tests/unit/file_handling_tests/test_fuzz_secret_file.py]: Seems that this is really doing your formatter. Here it as well tried to detect a comment and moved it one line up. Same as in the case with the link+anchor.

File: tests/unit/test_config_loading_and_dumping.py [hunk 12]
```diff
@@ -129,12 +153,61 @@ def test_load(
...
+
+    mock_hvac_client_read.assert_called_with(
+        gen_input_config()["vault_secrets"]["acme.user"]
+    )
+    mock_load.assert_called_with("in.json")
+
+
+@patch("vault_anyconfig.vault_anyconfig.load_base")
+@patch("vault_anyconfig.vault_anyconfig.Client.read")
+def test_load_kv2(
```
Comment (martinm82) [inline tests/unit/test_config_loading_and_dumping.py]: Feels like you could write parameterized tests and run all test cases once for kv1 and once for kv2. I guess that this test case does differ only in the `gen_vault_response_kvX` method call, right?

File: vault_anyconfig/vault_anyconfig.py [hunk 13]
```diff
@@ -274,6 +278,50 @@ def __process_vault_files(self, config):
...
+                "Invalid response recieved. Possibly due to an unsupported secrets engine, vault-anyconfig currently only supports kv1 and kv2.")
+        return secret_string
+
+    @staticmethod
+    def __is_key_value_v1(read_response, secret_key):
+        """
+        Checks if the response is from the key value v1 secret engine.
+        See https://www.vaultproject.io/api/secret/kv/kv-v1.html#sample-response-1
+        Args:
+            - read_response: response from HVAC read function
```
Comment (martinm82) [inline vault_anyconfig/vault_anyconfig.py]: Missing arg `secret_key`

File: vault_anyconfig/vault_anyconfig.py [hunk 14]
```diff
@@ -274,6 +278,50 @@ def __process_vault_files(self, config):
...
+        """
+        Detects the secret engine returning a secret (currently *only* supports key-value versions 1 and 2) and returns the requested key from the
+        secret.
+        Args:
+            - read_response: response from HVAC read function
+            - secret_key: secret key being retrieved
+        Returns:
+            secret string
+        """
+        if cls.__is_key_value_v1(read_response, secret_key):
```
Comment (martinm82) [inline vault_anyconfig/vault_anyconfig.py]: I thought you want to check for presence of the `metadata` key?

File: tests/unit/test_engine_detection.py [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+@example(contents="aoeu", secret_key="data")
+@example(contents="aoeu", secret_key="metadata")
+def test_detect_kv_v1(contents, secret_key, gen_vault_response_kv1, gen_vault_response_kv2):
+    """
+    Tests that kv1 is detected correctly
+    """
+    read_response_v1 = gen_vault_response_kv1(contents, secret_key)
+    read_response_v2 = gen_vault_response_kv2(contents, secret_key)
+
+    assert VaultAnyConfig._VaultAnyConfig__is_key_value_v1(
```
Comment (martinm82) [inline tests/unit/test_engine_detection.py]: What is this for a funny construct (`._VaultAnyConfig__is_key_value_v1`)?

Comment (eugene-davis) [reply tests/unit/test_engine_detection.py]: It allows accessing this internal function. Normally I would avoid testing it directly at all since its purely internal, but given the amount of difficulty in getting it working, I opted to test directly.

File: vault_anyconfig/vault_anyconfig.py (hunk 14 above)
Comment (eugene-davis) [reply vault_anyconfig/vault_anyconfig.py]: Originally yes, but since it isn't actually the primary differentiating aspect, I didn't include it in the final version.

File: vault_anyconfig/vault_anyconfig.py (hunk 13 above)
Comment (eugene-davis) [reply vault_anyconfig/vault_anyconfig.py]: Fixed

File: tests/unit/test_config_loading_and_dumping.py (hunk 12 above)
Comment (eugene-davis) [reply tests/unit/test_config_loading_and_dumping.py]: No, it also changes what is passed into the functions, since I check to make sure data from v1 will not trigger on the v2 detect and vice versa. Still possible to parameterize, but it will make it less clear what the test is actually testing, and I would rather avoid that.

File: README.md (hunk 10 above)
Comment (eugene-davis) [reply README.md]: Exactly, which will surely happen.

File: tests/unit/file_handling_tests/test_fuzz_secret_file.py (hunk 11 above)
Comment (eugene-davis) [reply tests/unit/file_handling_tests/test_fuzz_secret_file.py]: Good catch. Turning it off for the time being.

Comment (eugene-davis) [reply tests/unit/file_handling_tests/test_fuzz_secret_file.py]: Although in this case maybe it was a line-length issue, as this comment is a bit long
